src/cupbearer/detectors/finetuning.py

```diff
@@ -0,0 +1,145 @@
+
+            loss, grads = jax.value_and_grad(loss_fn)(state.params)
+            state = state.apply_gradients(grads=grads)
+            metrics = {"loss": loss}
+            return state, metrics
+
+        def eval_step(state, batch):
+            loss = self.loss_fn(state.params, batch)
+            metrics = {"loss": loss}
+            return metrics
+
+        return train_step, eval_step
+
+
+class FinetuningAnomalyDetector(AnomalyDetector):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.finetuned_params = None
+
+    def train(
+        self,
+        clean_dataset,
+        optimizer: OptimizerConfig,
+        num_epochs: int = 10,
+        batch_size: int = 128,
+        # Not used, but will be passed by the detector train script
+        debug: bool = False,
+    ):
+        trainer_instance = FinetuningTrainer(
+            model=self.model,
+            loss_fn=self.loss_fn,
+            optimizer=optimizer.build(),
+            log_dir=self.save_path,
+            override_variables={
+                "params": self.params
+            },  # Use existing parameters for initialization
+            example_input=clean_dataset[0][0][None],
+        )
+
+        # Create a DataLoader for the clean dataset
+        clean_loader = DataLoader(
+            dataset=clean_dataset,
+            batch_size=batch_size,
+            collate_fn=numpy_collate,
+        )
+
+        # Finetune the model on the clean dataset
+        trainer_instance.train_model(
+            train_loader=clean_loader,
+            val_loaders={},
+            num_epochs=num_epochs,
+        )
+
+        # Store the finetuned parameters
+        self.finetuned_params = trainer_instance.state.params
+
+    def loss_fn(self, params, batch):
+        inputs, targets = batch
+        logits = self.model.apply({"params": params}, inputs)
+        return optax.softmax_cross_entropy_with_integer_labels(logits, targets).mean()
+
+    def layerwise_scores(self, batch):
+        raise NotImplementedError(
+            "Layerwise scores don't exist for finetuning detector"
+        )
+
+    def scores(self, batch):
+        original_output, _ = self._model(batch)
+
+        if isinstance(batch, (tuple, list)):
+            inputs = batch[0]
+        else:
+            inputs = batch
+        finetuned_output = self.model.apply({"params": self.finetuned_params}, inputs)
+
+        finetuned_p = jax.nn.softmax(finetuned_output, axis=-1)
+        original_p = jax.nn.softmax(original_output, axis=-1)
+
+        # Check whether we're going to get infinities:
+        if jnp.any(jnp.logical_and(finetuned_p == 0, original_p > 0)):
+            # We'd get an error anyway once we compute eval metrics, better to give
+            # a more specific one here.
+            raise ValueError("Infinite KL divergence")
+
+        # TODO: which direction do we want here? Or maybe just some other metric?
```

Comment:
I'd say we should consider the finetuned distribution to be the "ground truth" as we're comparing to see if we're surprised by the original output. So this would be D(finetuned_p||original_p) which (checking rel_entr docs) would be the other way around.

I'm unsure about this so please convince me otherwise.

Reply:
Hm, I feel like this is a bit unprincipled anyway (it's not clear why there should be any asymmetry). So that points towards some other metric. But I checked some previous work by Redwood and they're using KL divergence with D(original || finetuned) apparently. I think for now let's just stick with that, I'll update the comment.

Empirically, it seems that everything works similarly well on a very quick test (including weird stuff like an L2 distance between probability distributions). If we think the metric is important, we should actually carefully compare different options at some point, but that's pretty far down my list of things to do for now. Might be more appealing if this baseline looks to be the strongest one after a while.